### PR TITLE
feat: Add support for Kafka clusters sharding

### PR DESCRIPTION
**Pull request by olksdr**
This allows to configured many different Kafka clusters per topic. The
basic idea is to provide the numebr of shards and the lower bound for
the range with the kafka config name and the topic name on that cluster

```
 metrics:
     shards: 65000
     mapping:
         0:
             name: "ingest-metrics-1"
             config: "metric_1"
         25000:
             name: "ingest-metrics-2"
             config: "metrics_2"
         45000:
             name: "ingest-metrics-3"
             config: "metrics_3"
```

This will also allow to configure one Kafka clusters and use different
topics names on it.

### CODE REVIEW

File: relay-config/src/config.rs [hunk 1]
```diff
@@ -882,41 +885,106 @@ pub enum TopicAssignment {
...
+    #[serde(rename = "config")]
+    kafka_config_name: String,
+}
+
+/// Configuration for logical shards -> kafka configuration mapping.
+#[derive(Serialize, Deserialize, Debug)]
+pub struct Sharded {
+    /// The number of shards used for this topic.
+    shards: u64,
+    /// The Kafka configuration assigned to the specific shard range.
```
Comment (flub) [inline relay-config/src/config.rs:932]: could you describe that the u64 is the start index and the next u64 describes the range.  Explicitly calling out what's inclusive (i'm assuming the start u64 is inclusive, the end u64 is excluded and part of the next range).  Maybe even write an example out on the struct doc comment like you did in the PR description.

File: relay-server/src/actors/outcome.rs [hunk 2]
```diff
@@ -835,17 +820,46 @@ impl OutcomeProducerService {
...
+                producer.send(record)
+            }
+
+            Producer::Sharded { shards, producers } => {
+                let mut hasher = FnvHasher::default();
+                std::hash::Hash::hash(&organization_id, &mut hasher);
+                let org_id_hash = hasher.finish();
+                let shard = org_id_hash % shards;
+
+                // should be ok to unwrap since we MUST have at least one range defined
```
Comment (flub) [inline relay-server/src/actors/outcome.rs]: This is reasonable but because it is so far apart from each other in the code I'm a bit uncomfortable about it.  One way would be make a custom type for the producers, then it is enforced in that type and relying on that invariant would be a lot nicer and keeps the code interacting with the invariant close together.  I guess this would be a newtype over the BTreeMap with a few small methods.

File: relay-server/src/actors/store.rs [hunk 3]
```diff
@@ -58,33 +58,127 @@ fn make_distinct_id(s: &str) -> Uuid {
...
+                producer
+                    .send(record)
+                    .map_err(|(kafka_error, _message)| StoreError::SendFailed(kafka_error))?;
+            }
+
+            Self::Sharded { shards, producers } => {
+                let mut hasher = FnvHasher::default();
+                std::hash::Hash::hash(&organization_id, &mut hasher);
+                let org_id_hash = hasher.finish();
+                let shard = org_id_hash % shards;
```
Comment (flub) [inline relay-server/src/actors/store.rs]: this code exists in two places already, given the nature of it - i suspect both would need updating together all the time - it might be better to create a method somewhere.  Maybe also on the newtype around the BTreeMap I already suggested.

**Review by flub — APPROVED** [review]

File: relay-server/src/actors/outcome.rs (hunk 2 above)
Comment (olksdr) [reply relay-server/src/actors/outcome.rs]: That's a very good point. I will look into it. 
It might also simplify the code and removes some duplication.

File: relay-config/src/config.rs (hunk 1 above)
Comment (olksdr) [reply relay-config/src/config.rs:932]: @flub, please, have a look into https://github.com/getsentry/relay/pull/1454/commits/f9679331735dcfa7c8a490bcc6922e69f2a0d0f9 if this is something you had in mind?

File: tests/integration/fixtures/relay.py [hunk 4]
```diff
@@ -133,7 +134,7 @@ def inner(
                 default_opts.setdefault(key, {}).update(options[key])
 
         dir = config_dir("relay")
-        dir.join("config.yml").write(json.dumps(default_opts))
+        dir.join("config.yml").write(yaml.dump(default_opts))
```
Comment (jjbayer) [inline tests/integration/fixtures/relay.py:137]: Why is this change necessary?

File: relay-server/src/actors/outcome.rs (hunk 2 above)
Comment (olksdr) [reply relay-server/src/actors/outcome.rs]: So, I moved the same functionality under the `Sharded` producer in store actore [d4a05c2](https://github.com/getsentry/relay/pull/1454/commits/d4a05c23f39c9238481d9338b0b6b7f73ebb7010). 
I do not know if it's a good idea to create a new type over `BTreeMap`, which can be a bit more complicated to understand.
But I might have to pick your brain on this , and how to implement better.

File: relay-server/src/actors/store.rs (hunk 3 above)
Comment (olksdr) [reply relay-server/src/actors/store.rs]: moved this into the impl for `Sharded` producer in  [d4a05c2](https://github.com/getsentry/relay/pull/1454/commits/d4a05c23f39c9238481d9338b0b6b7f73ebb7010)

File: relay-server/src/actors/store.rs [hunk 5]
```diff
@@ -58,33 +58,143 @@ fn make_distinct_id(s: &str) -> Uuid {
...
+            }
+
+            Self::Sharded(sharded) => {
+                let (topic_name, producer) = sharded.get_producer(organization_id);
+                let record = BaseRecord::to(topic_name).key(&key).payload(&serialized);
+                producer
+                    .send(record)
+                    .map_err(|(kafka_error, _message)| StoreError::SendFailed(kafka_error))?;
+            }
+        }
```
Comment (jjbayer) [inline relay-server/src/actors/store.rs]: There's code duplication in here, might be nice to only use the `match` statement to obtain the topic name & producer:

```suggestion
        let (topic_name, producer) = match self {
            Producer::Single {
                topic_name,
                producer,
            } => (topic_name.as_str(), producer),
            Producer::Sharded(sharded) => sharded.get_producer(organization_id),
        };

        let record = BaseRecord::to(topic_name).key(&key).payload(&serialized);
        producer
            .send(record)
            .map_err(|(kafka_error, _message)| StoreError::SendFailed(kafka_error))?;

```

File: relay-server/src/actors/store.rs [hunk 6]
```diff
@@ -58,33 +58,143 @@ fn make_distinct_id(s: &str) -> Uuid {
...
+        topic_name: String,
+        producer: Arc<ThreadedProducer>,
+    },
+    Sharded(Sharded),
+}
 
-    for config_p in kafka_config {
-        client_config.set(config_p.name.as_str(), config_p.value.as_str());
+/// Sharded producer configuration.
+pub struct Sharded {
```
Comment (jjbayer) [inline relay-server/src/actors/store.rs]: nit: Should we call this `ShardedProducer`?

File: relay-server/src/actors/store.rs [hunk 7]
```diff
@@ -58,33 +58,143 @@ fn make_distinct_id(s: &str) -> Uuid {
...
+    /// The maximum number of shards for this producer.
+    shards: u64,
+    /// The actual Kafka producer assigned to the range of logical shards, where the `u64` in the map is
+    /// the inclusive beginning of the range.
+    producers: BTreeMap<u64, (String, Arc<ThreadedProducer>)>,
+}
+
+impl Sharded {
+    /// Returns topic name and the Kafka producer based on the provided sharding key.
+    pub fn get_producer(&self, sharding_key: u64) -> (&str, Arc<ThreadedProducer>) {
```
Comment (jjbayer) [inline relay-server/src/actors/store.rs]: Might be a premature generalization, but we could make this function generic:
```suggestion
    pub fn get_producer(&self, sharding_key: impl Hash) -> (&str, Arc<ThreadedProducer>) {
```

File: relay-server/src/actors/store.rs [hunk 8]
```diff
@@ -58,33 +58,143 @@ fn make_distinct_id(s: &str) -> Uuid {
...
+        let mut hasher = FnvHasher::default();
+        std::hash::Hash::hash(&sharding_key, &mut hasher);
+        let org_id_hash = hasher.finish();
+        let shard = org_id_hash % self.shards;
```
Comment (jjbayer) [inline relay-server/src/actors/store.rs]: Since `org_id` is not in the function signature, I would also rename this variable:
```suggestion
        let hashed = hasher.finish();
        let shard = hashed % self.shards;
```

File: relay-config/src/config.rs [hunk 9]
```diff
@@ -882,41 +885,126 @@ pub enum TopicAssignment {
...
+pub enum KafkaConfig<'a> {
+    /// Single config with Kafka parameters.
+    Single(KafkaParams<'a>),
+
+    /// The list of the Kafka configs with related shard configs.
+    Sharded {
+        /// The maximum number of logical shards for this set of configs.
+        shards: u64,
+        /// The list of the sharded Kafka configs.
+        configs: BTreeMap<u64, KafkaParams<'a>>,
```
Comment (jjbayer) [inline relay-config/src/config.rs:948]: We have this mapping in three different places now, I wonder if it would be possible to have it only once.

File: tests/integration/fixtures/relay.py (hunk 4 above)
Comment (olksdr) [reply tests/integration/fixtures/relay.py:137]: There are few things here:
* first of all I think we should use Yaml for testing the configuration, since we use Yaml config file to configure the service (not JSON) - so we can test this as closer to real world 
* secondly, and which is more relevant to this change - that with new additions to the configuration we have to have the integers as the keys in the map, but JSON can only have strings there. 

And here is the thing, we if want to also support JSON we will have to relay on parsing the strings, maybe in addition to numbers (?),  and getting integers from there.  What do you think?

File: relay-server/src/actors/store.rs (hunk 6 above)
Comment (olksdr) [reply relay-server/src/actors/store.rs]: Yeah, this is must better name!  I will make a change.

File: tests/integration/fixtures/relay.py (hunk 4 above)
Comment (jjbayer) [reply tests/integration/fixtures/relay.py:137]: The integer keys make sense. To maintain JSON compatibility, we could also change the format to a list (like below), but then we would have to validate that there are no duplicate entries. So I think sticking with yaml makes sense.

```yaml
ranges:
    - start_at: 0
      name: "ingest-metrics-1"
      config: "metrics_1"
    - start_at: 25000
      name: "ingest-metrics-2"
      config: "metrics_2"
```

**Review by iker-barriocanal — APPROVED** [review]

File: relay-config/src/config.rs (hunk 1 above)
Comment (flub) [reply relay-config/src/config.rs:932]: yep, 💯

File: relay-config/src/config.rs (hunk 9 above)
Comment (olksdr) [reply relay-config/src/config.rs:948]: @jjbayer I've tried to remove one of the levels of indirection to eliminate of of the repeating BTreeMap, see [0db194e](https://github.com/getsentry/relay/pull/1454/commits/0db194e5b2a53a2cd8462d8dc369c997a38f6439) 
Please, have another look if this is somewhat better approach.

File: relay-config/src/config.rs [hunk 10]
```diff
@@ -2236,11 +2200,5 @@ processing:
             };
         assert_eq!(shards, 65000);
         assert_eq!(3, mapping.len());
-
-        let kafka_config_profiles = values.processing.topics.events.kafka_config_name();
-        assert!(matches!(
-            kafka_config_profiles,
-            KafkaConfigName::Single { .. }
-        ));
```
Comment (jjbayer) [inline relay-config/src/config.rs]: Any way we could keep this assertion?

File: relay-config/src/config.rs [hunk 11]
```diff
@@ -985,27 +968,56 @@ impl From<String> for TopicAssignment {
 
 impl TopicAssignment {
     /// Get the name of the kafka config to use. `None` means default configuration.
-    fn kafka_config_name(&self) -> KafkaConfigName {
-        match self {
-            Self::Primary(topic) => KafkaConfigName::Single {
-                topic_name: topic.as_str(),
-                kafka_config_name: None,
-            },
+    fn kafka_config_name<'a>(
```
Comment (jjbayer) [inline relay-config/src/config.rs]: Should we update the doc comment and the function name, now that this function returns a full config rather than a name?

File: relay-config/src/config.rs (hunk 9 above)
Comment (jjbayer) [reply relay-config/src/config.rs:948]: Yes, I think it makes sense this way.

File: relay-server/src/actors/store.rs [hunk 12]
```diff
@@ -58,33 +58,134 @@ fn make_distinct_id(s: &str) -> Uuid {
...
+        metric!(
+            histogram(RelayHistograms::KafkaMessageSize) = serialized.len() as u64,
+            variant = message.variant()
+        );
+        let key = message.key();
+        let (topic_name, producer) = match self {
+            Self::Single {
+                topic_name,
+                producer,
+            } => (topic_name.as_str(), Arc::clone(producer)),
```
Comment (jjbayer) [inline relay-server/src/actors/store.rs]: nit: You can probably change the signature of `get_producer` to ` return (&str, &ThreadedProducer)` to save on `Arc::clone`s.

**Review by jjbayer — APPROVED** [review]

File: tests/integration/fixtures/relay.py (hunk 4 above)
Comment (untitaker) [inline tests/integration/fixtures/relay.py:137]: we don't have to change this now, but I believe that at some point the configuration will have to become json-serializable. then we will have to revisit the schema. but that's fine

**Review by untitaker — APPROVED** [review]

File: relay-config/src/config.rs (hunk 11 above)
Comment (olksdr) [reply relay-config/src/config.rs]: Good point, done in [1c3d5e0](https://github.com/getsentry/relay/pull/1454/commits/1c3d5e07e5a6d8352b6fadbed57a04d649374f0f)

File: relay-config/src/config.rs (hunk 10 above)
Comment (olksdr) [reply relay-config/src/config.rs]: done in [1c3d5e0](https://github.com/getsentry/relay/pull/1454/commits/1c3d5e07e5a6d8352b6fadbed57a04d649374f0f).

File: relay-server/src/actors/store.rs (hunk 12 above)
Comment (olksdr) [reply relay-server/src/actors/store.rs]: done in [1c3d5e0](https://github.com/getsentry/relay/pull/1454/commits/1c3d5e07e5a6d8352b6fadbed57a04d649374f0f)
